proxy_pn_counter.go

```diff
@@ -0,0 +1,227 @@
+aren't any and the methods here are invoked on a lite member, they will
+fail with an NoDataMemberInClusterError.
+*/
+type PNCounter struct {
+	*proxy
+	clock  iproxy.VectorClock
+	target atomic.Value
+	mu     *sync.Mutex
+}
+
+func newPNCounter(p *proxy) *PNCounter {
+	return &PNCounter{
+		proxy: p,
+		clock: iproxy.NewVectorClock(),
+		mu:    &sync.Mutex{},
+	}
+}
+
+// AddAndGet adds the given value to the current value and returns the updated value.
+func (pn *PNCounter) AddAndGet(ctx context.Context, delta int64) (int64, error) {
+	return pn.add(ctx, delta, false)
+}
+
+// DecrementAndGet decrements the counter value by one and returns the updated value.
+func (pn *PNCounter) DecrementAndGet(ctx context.Context) (int64, error) {
+	return pn.add(ctx, -1, false)
+}
+
+// Get returns the current value of the counter.
+func (pn *PNCounter) Get(ctx context.Context) (int64, error) {
+	resp, err := pn.invokeOnTarget(ctx, func(uuid types.UUID) *proto.ClientMessage {
+		return codec.EncodePNCounterGetRequest(pn.name, pn.clockEntrySet(), uuid)
+	})
+	if err != nil {
+		return 0, err
+	}
+	value, timestamps, _ := codec.DecodePNCounterGetResponse(resp)
+	pn.updateClock(iproxy.NewVectorClockFromPairs(timestamps))
+	return value, nil
+}
+
+// GetAndAdd adds the given value to the current value and returns the previous value.
+func (pn *PNCounter) GetAndAdd(ctx context.Context, delta int64) (int64, error) {
+	return pn.add(ctx, delta, true)
+}
+
+// GetAndDecrement decrements the counter value by one and returns the previous value.
+func (pn *PNCounter) GetAndDecrement(ctx context.Context) (int64, error) {
+	return pn.add(ctx, -1, true)
+}
+
+// GetAndIncrement increments the counter value by one and returns the previous value.
+func (pn *PNCounter) GetAndIncrement(ctx context.Context) (int64, error) {
+	return pn.add(ctx, 1, true)
+}
+
+// GetAndSubtract subtracts the given value from the current value and returns the previous value.
+func (pn *PNCounter) GetAndSubtract(ctx context.Context, delta int64) (int64, error) {
+	return pn.add(ctx, -1*delta, true)
+}
+
+// IncrementAndGet increments the counter value by one and returns the updated value.
+func (pn *PNCounter) IncrementAndGet(ctx context.Context) (int64, error) {
+	return pn.add(ctx, 1, false)
+}
+
+// Reset resets the observed state by this PN counter.
+func (pn *PNCounter) Reset() {
+	pn.mu.Lock()
+	pn.clock = iproxy.NewVectorClock()
+	pn.mu.Unlock()
+}
+
+// SubtractAndGet subtracts the given value from the current value and returns the updated value.
+func (pn *PNCounter) SubtractAndGet(ctx context.Context, delta int64) (int64, error) {
+	return pn.add(ctx, -1*delta, false)
+}
+
+func (pn *PNCounter) clockEntrySet() []proto.Pair {
+	pn.mu.Lock()
+	entries := pn.clock.EntrySet()
+	pn.mu.Unlock()
+	return entries
+}
+
+func (pn *PNCounter) crdtOperationTarget(excluded map[cluster.Address]struct{}) (*cluster.MemberInfo, error) {
+	target := pn.target.Load()
```

Comment:
It looks like two concurrent callers of this method my get two different members as return value. I think we need more synchronization here.

Reply:
Fixed at: https://github.com/hazelcast/hazelcast-go-client/pull/579/commits/34cf0b6e9351b5b899f65d27f596ce2a060aa5a5